File: haskell/glue/src/Glue/Lib/Math/Power/Exp.hs
module Glue.Lib.Math.Power.Exp where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Exponential function (e^x)
-- Mirrors Haskell Glue.Lib.Math.Power.Exp.exp exactly
exp :: IR Eval
exp = NativeFunc expImpl

-- Exponential function implementation
-- Mirrors Haskell Glue.Lib.Math.Power.Exp.expImpl exactly
expImpl :: IR Eval -> Eval (IR Eval)
expImpl arg = case arg of
    Integer n -> pure $ Float (Prelude.exp (fromIntegral n))
    Float n -> pure $ Float (Prelude.exp n)
    _ -> throwError $ wrongArgumentType ["number"]

File: dart/glue/lib/src/lib/math/power/exp.dart
import 'dart:math' as math;

import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Exponential function (e^x)
/// Mirrors Haskell Glue.Lib.Math.Power.Exp.exp exactly
final Ir exp = IrNativeFunc(expImpl);

/// Exponential function implementation (e^x)
/// Mirrors Haskell Glue.Lib.Math.Power.Exp.expImpl exactly
Eval<Ir> expImpl(Ir arg) {
  return switch (arg) {
    IrInteger(value: final n) => Eval.pure(IrFloat(math.exp(n.toDouble()))),
    IrFloat(value: final n) => Eval.pure(IrFloat(math.exp(n))),
    _ => throwError(wrongArgumentType(['number'])),
  };
}
